File: src/tests/py_unittest/packet_conv/Test_volume_modify.py
# coding: UTF-8
"""Tests for ``packet_conv/volume_modify.py``

Test cases:
  * Can encode/decode volume modify packet
  * Raise TypeError if invalid argument is passed
  * Raise ValueError if invalid argument is passed

Test steps:
  * Run this script by unittest - ``python3 -m unittest discover "Test_*.py"``

Note:
  Specification of ``volume_modify`` packet, see ``/designs/packet-protocols.md``

"""

# For import top layer module
import sys
from pathlib import Path
sys.path.append(Path(__file__).absolute().parent.parent.parent.__str__())


import unittest

from uuid import uuid4

from modules.packet_conv import volume_modify


class Test_packet_conv_volume_modify(unittest.TestCase):
  def test_true_enc_dec(self):
    lane_id_org = 1
    modified_volume_org = 100
    raw_packet = volume_modify.encode(lane_id_org, modified_volume_org)
    lane_id_prc, modified_volume_prc = volume_modify.decode(raw_packet)

    self.assertEqual(lane_id_org, lane_id_prc)
    self.assertEqual(modified_volume_org, modified_volume_prc)


  def test_err_enc_invalid_type(self):
    self.assertRaises(TypeError, volume_modify.encode, "", 0) # str "" as non int
    self.assertRaises(TypeError, volume_modify.encode, 0, "") # str "" as non int


  def test_err_enc_invalid_value(self):
    self.assertRaises(ValueError, volume_modify.encode, -1 , 0) # under range
    self.assertRaises(ValueError, volume_modify.encode, 256, 0) # over range
    self.assertRaises(ValueError, volume_modify.encode, 0, -1)  # under range
    self.assertRaises(ValueError, volume_modify.encode, 0, 256) # over range


  def test_err_dec_invalid_type(self):
    self.assertRaises(TypeError, volume_modify.decode, "")  # str "" as non bytes


  def test_err_dec_invalid_value(self):
    raw_packet_invalid_id = b"A" + bytes([1, 2])
    #                       ~~~~ as non 0x20 byte
    raw_packet_invalid_len = bytes([volume_modify.VOLUME_MODIFY_PACKET_TYPE_ID, 1])
    # modified volume missing packet

    self.assertRaises(ValueError, volume_modify.decode, raw_packet_invalid_id)
    self.assertRaises(ValueError, volume_modify.decode, raw_packet_invalid_len)


if __name__ == "__main__":
  unittest.main(verbosity=2)

File: src/tests/js_vitest/packet_conv/Test_volume_modify.js
/**
 * @file Tests for `packet_conv/volume_modify.js`
 *
 * Test cases:
 *   * Can encode/decode audio packet
 *   * Raise TypeError if invalid argument type is passed
 *   * Raise RangeError if invalid argument value is passed
 *
 * Test steps:
 *   * Run this script by vitest - `npm test`
 *
 * @author aKuad
 */

import { packet_volume_modify_encode, packet_volume_modify_decode, is_volume_modify_packet } from "../../../static/packet_conv/volume_modify.js";
import { describe, test, expect } from "vitest";


describe("true_cases", () => {
  test("enc_dec_verify", () => {
    const lane_id_org = 1;
    const modified_volume_org = 100;
    const raw_packet = packet_volume_modify_encode(lane_id_org, modified_volume_org);
    const [lane_id_prc, modified_volume_prc] = packet_volume_modify_decode(raw_packet);

    expect(is_volume_modify_packet(raw_packet)).toBe(true);
    expect(lane_id_prc).toBe(lane_id_org);
    expect(modified_volume_prc).toBe(modified_volume_org);
  });


  test("verify_ng", () => {
    const lane_id_org = 1;
    const modified_volume_org = 100;
    const raw_packet_correct = packet_volume_modify_encode(lane_id_org, modified_volume_org);

    const raw_packet_invalid_empty     = new Uint8Array();
    const raw_packet_invalid_id        = Uint8Array.of(0x30, ...raw_packet_correct.slice(1));  // as non 0x20 byte
    const raw_packet_invalid_too_short = raw_packet_correct.slice(0, -1);
    const raw_packet_invalid_too_long  = Uint8Array.of(...raw_packet_correct, 0); // 0 as an over length byte

    expect(is_volume_modify_packet("")                          ).toBe(false);  // string "" as non Uint8Array
    expect(is_volume_modify_packet(raw_packet_invalid_empty)    ).toBe(false);
    expect(is_volume_modify_packet(raw_packet_invalid_id)       ).toBe(false);
    expect(is_volume_modify_packet(raw_packet_invalid_too_short)).toBe(false);
    expect(is_volume_modify_packet(raw_packet_invalid_too_long) ).toBe(false);

    expect(() => is_volume_modify_packet(""                          , true)).toThrowError(new TypeError ("raw_packet must be Uint8Array"));
    expect(() => is_volume_modify_packet(raw_packet_invalid_empty    , true)).toThrowError(new RangeError("Empty array passed"));
    expect(() => is_volume_modify_packet(raw_packet_invalid_id       , true)).toThrowError(new RangeError("It is not a volume_modify packet"));
    expect(() => is_volume_modify_packet(raw_packet_invalid_too_short, true)).toThrowError(new RangeError("Too short bytes as volume modify packet"));
    expect(() => is_volume_modify_packet(raw_packet_invalid_too_long , true)).toThrowError(new RangeError("Too long bytes as volume modify packet"));
  });
});


describe("err_cases", () => {
  test("enc_invalid_type", () => {
    const lane_id = 1;
    const modified_volume = 100;

    expect(() => packet_volume_modify_encode("", modified_volume)).toThrowError(new TypeError("lane_id must be number"));
    expect(() => packet_volume_modify_encode(lane_id, ""        )).toThrowError(new TypeError("modified_volume must be number"));
  });


  test("enc_invalid_value", () => {
    const lane_id = 1;
    const modified_volume = 100;

    expect(() => packet_volume_modify_encode(256, modified_volume)).toThrowError(new RangeError("lane_id must be in 0~255"));
    expect(() => packet_volume_modify_encode(-1 , modified_volume)).toThrowError(new RangeError("lane_id must be in 0~255"));
    expect(() => packet_volume_modify_encode(lane_id, 256        )).toThrowError(new RangeError("modified_volume must be in 0~255"));
    expect(() => packet_volume_modify_encode(lane_id, -1         )).toThrowError(new RangeError("modified_volume must be in 0~255"));
  });


  // test("dec_invalid_type", () => {
  //   // packet verification integrated to `is_volume_modify_packet` tests
  // });

  // test("dec_invalid_value", () => {
  //   // packet verification integrated to `is_volume_modify_packet` tests
  // });

  // test("verify_invalid_type", () => {
  //   // no error cases of `is_volume_modify_packet`
  // });

  // test("verify_invalid_value", () => {
  //   // no error cases of `is_volume_modify_packet`
  // });
});
